testing/e2e/tests/test.go

```diff
@@ -0,0 +1,67 @@
+package tests
+
+import (
+	"context"
+
+	"github.com/Azure/aks-app-routing-operator/testing/e2e/infra"
+	"github.com/Azure/aks-app-routing-operator/testing/e2e/logger"
+	"golang.org/x/sync/errgroup"
+)
+
+func (tests Ts) Run(ctx context.Context, infra infra.Provisioned) error {
+	var parallel []T
+	var sequential []T
+
+	lgr := logger.FromContext(ctx)
+
+	// divide tests according to run strategy and filter out tests that should not run
+	for _, test := range tests {
+		if !test.ShouldRun(infra) {
+			lgr.Info("skipping test", "test", test.GetName())
+		}
+
+		if test.GetRunStrategy() == Parallel {
+			parallel = append(parallel, test)
+		} else {
+			sequential = append(sequential, test)
+		}
+	}
+
+	runTestFn := func(t T) *logger.LoggedError {
+		lgr := lgr.With("test", t.GetName())
+		ctx := logger.WithContext(ctx, lgr)
+		lgr.Info("starting to run test")
+
+		if err := t.Run(ctx); err != nil {
+			return logger.Error(lgr, err)
+		}
+
+		lgr.Info("finished running test")
+		return nil
+	}
+
+	lgr.Info("running sequential tests")
+	for _, test := range sequential {
```

Comment:
Can we define tests in such a way the parallel jobs can be kicked off first with sequential jobs starting at the same time? Or is this an infra restriction? Can you provide an example of what youre thinking that would block it

Reply:
It's an infra restriction. Sequential tests would be tests that can't be run at the same time as any other test.

An example sequential test would be testing the cleanup of external dns when no dns ids are passed in to the operator (we remove any running external dns deployments). That shouldn't run at the same time as other tests because it could impact them since other tests would rely on external dns.

Reply:
Yes. No reason to provision multiple of the same cluster unless there's a cluster level configuration option.

We deploy a few different kinds of clusters (osm enabled, large node pool, other cluster-level configurations) then run all tests on those clusters (unless that test wouldn't work on that cluster through the `ShouldRun(infra infra.Provisioned) bool` function). An example of a test not working on a certain kind of infra would be something exclusively testing osm that wouldn't work on non-osm clusters.